File: python/src/pi_mono/coding_agent/utils/tools_manager.py
"""Managed external tool discovery and download (fd, rg)."""

from __future__ import annotations

import os
import platform
import shutil
import subprocess
import sys
import tarfile
import tempfile
import time
import zipfile
from dataclasses import dataclass
from pathlib import Path
from typing import Callable
from urllib.request import Request, urlopen

from pi_mono.config import APP_NAME, get_bin_dir

ToolName = str

NETWORK_TIMEOUT_S = 10.0
DOWNLOAD_TIMEOUT_S = 120.0

_FD_SYSTEM_NAMES = ("fd", "fdfind")
_RG_SYSTEM_NAMES = ("rg",)

TERMUX_PACKAGES = {"fd": "fd", "rg": "ripgrep"}


@dataclass(frozen=True)
class ToolConfig:
    name: str
    repo: str
    binary_name: str
    system_binary_names: tuple[str, ...]
    tag_prefix: str
    get_asset_name: Callable[[str, str, str], str | None]


def _is_offline_mode_enabled() -> bool:
    value = os.environ.get("PI_OFFLINE", "")
    return value in ("1", "true", "True", "yes", "YES")


def _command_exists(command: str) -> bool:
    if shutil.which(command) is None:
        return False
    try:
        completed = subprocess.run(
            [command, "--version"],
            capture_output=True,
            check=False,
        )
        return completed.returncode == 0
    except OSError:
        return False


def _get_latest_version(repo: str) -> str:
    request = Request(
        f"https://api.github.com/repos/{repo}/releases/latest",
        headers={"User-Agent": f"{APP_NAME}-coding-agent"},
    )
    with urlopen(request, timeout=NETWORK_TIMEOUT_S) as response:
        import json

        data = json.loads(response.read().decode("utf-8"))
    tag = str(data.get("tag_name", ""))
    return tag.removeprefix("v")


def _download_file(url: str, dest: str) -> None:
    request = Request(url, headers={"User-Agent": f"{APP_NAME}-coding-agent"})
    with urlopen(request, timeout=DOWNLOAD_TIMEOUT_S) as response, open(dest, "wb") as handle:
        while True:
            chunk = response.read(1024 * 64)
            if not chunk:
                break
            handle.write(chunk)


def _find_binary_recursively(root_dir: str, binary_name: str) -> str | None:
    for root, _dirs, files in os.walk(root_dir):
        if binary_name in files:
            return os.path.join(root, binary_name)
    return None


def _extract_archive(archive_path: str, extract_dir: str) -> None:
    if archive_path.endswith(".tar.gz"):
        with tarfile.open(archive_path, "r:gz") as archive:
            archive.extractall(extract_dir)
        return
    if archive_path.endswith(".zip"):
        with zipfile.ZipFile(archive_path) as archive:
            archive.extractall(extract_dir)
        return
    raise RuntimeError(f"Unsupported archive format: {archive_path}")


TOOLS: dict[str, ToolConfig] = {
    "fd": ToolConfig(
        name="fd",
        repo="sharkdp/fd",
        binary_name="fd",
        system_binary_names=_FD_SYSTEM_NAMES,
        tag_prefix="v",
        get_asset_name=lambda version, plat, arch: (
            f"fd-v{version}-aarch64-apple-darwin.tar.gz"
            if plat == "darwin" and arch == "arm64"
            else (
                f"fd-v{version}-x86_64-apple-darwin.tar.gz"
                if plat == "darwin"
                else (
                    f"fd-v{version}-aarch64-unknown-linux-gnu.tar.gz"
                    if plat == "linux" and arch == "arm64"
                    else (
                        f"fd-v{version}-x86_64-unknown-linux-gnu.tar.gz"
                        if plat == "linux"
                        else (
                            f"fd-v{version}-aarch64-pc-windows-msvc.zip"
                            if plat == "win32" and arch == "arm64"
                            else (
                                f"fd-v{version}-x86_64-pc-windows-msvc.zip"
                                if plat == "win32"
                                else None
                            )
                        )
                    )
                )
            )
        ),
    ),
    "rg": ToolConfig(
        name="ripgrep",
        repo="BurntSushi/ripgrep",
        binary_name="rg",
        system_binary_names=_RG_SYSTEM_NAMES,
        tag_prefix="",
        get_asset_name=lambda version, plat, arch: (
            f"ripgrep-{version}-aarch64-apple-darwin.tar.gz"
            if plat == "darwin" and arch == "arm64"
            else (
                f"ripgrep-{version}-x86_64-apple-darwin.tar.gz"
                if plat == "darwin"
                else (
                    f"ripgrep-{version}-aarch64-unknown-linux-gnu.tar.gz"
                    if plat == "linux" and arch == "arm64"
                    else (
                        f"ripgrep-{version}-x86_64-unknown-linux-musl.tar.gz"
                        if plat == "linux"
                        else (
                            f"ripgrep-{version}-aarch64-pc-windows-msvc.zip"
                            if plat == "win32" and arch == "arm64"
                            else (
                                f"ripgrep-{version}-x86_64-pc-windows-msvc.zip"
                                if plat == "win32"
                                else None
                            )
                        )
                    )
                )
            )
        ),
    ),
}


def get_tool_path(tool: ToolName) -> str | None:
    """Return a local or system path for fd/rg, or None if unavailable."""
    config = TOOLS.get(tool)
    if config is None:
        return None

    bin_dir = str(get_bin_dir())
    binary_ext = ".exe" if sys.platform == "win32" else ""
    local_path = os.path.join(bin_dir, config.binary_name + binary_ext)
    if os.path.isfile(local_path):
        return local_path

    for name in config.system_binary_names:
        if _command_exists(name):
            return name
    return None


async def _download_tool(tool: ToolName) -> str:
    config = TOOLS.get(tool)
    if config is None:
        raise RuntimeError(f"Unknown tool: {tool}")

    plat = sys.platform
    machine = platform.machine().lower()
    arch = "arm64" if machine in ("arm64", "aarch64") else "x86_64"

    version = await __import__("asyncio").to_thread(_get_latest_version, config.repo)
    if tool == "fd" and plat == "darwin" and arch == "x86_64":
        version = "10.3.0"

    asset_name = config.get_asset_name(version, plat, arch)
    if not asset_name:
        raise RuntimeError(f"Unsupported platform: {plat}/{arch}")

    bin_dir = Path(get_bin_dir())
    bin_dir.mkdir(parents=True, exist_ok=True)
    binary_ext = ".exe" if plat == "win32" else ""
    binary_path = bin_dir / f"{config.binary_name}{binary_ext}"

    download_url = (
        f"https://github.com/{config.repo}/releases/download/"
        f"{config.tag_prefix}{version}/{asset_name}"
    )
    archive_path = bin_dir / asset_name
    await __import__("asyncio").to_thread(_download_file, download_url, str(archive_path))

    extract_dir = tempfile.mkdtemp(
        prefix=f"extract_{config.binary_name}_{os.getpid()}_{int(time.time())}_",
        dir=str(bin_dir),
    )
    try:
        await __import__("asyncio").to_thread(_extract_archive, str(archive_path), extract_dir)
        binary_file_name = config.binary_name + binary_ext
        extracted_binary = _find_binary_recursively(extract_dir, binary_file_name)
        if not extracted_binary:
            raise RuntimeError(f"Binary not found in archive: {binary_file_name}")
        shutil.move(extracted_binary, binary_path)
        if plat != "win32":
            os.chmod(binary_path, 0o755)
    finally:
        archive_path.unlink(missing_ok=True)
        shutil.rmtree(extract_dir, ignore_errors=True)

    return str(binary_path)


async def ensure_tool(tool: ToolName, silent: bool = False) -> str | None:
    """Ensure fd/rg is available, downloading from GitHub releases when needed."""
    existing = get_tool_path(tool)
    if existing:
        return existing

    config = TOOLS.get(tool)
    if config is None:
        return None

    if _is_offline_mode_enabled():
        if not silent:
            print(f"{config.name} not found. Offline mode enabled, skipping download.")
        return None

    if sys.platform == "android":
        pkg_name = TERMUX_PACKAGES.get(tool, tool)
        if not silent:
            print(f"{config.name} not found. Install with: pkg install {pkg_name}")
        return None

    if not silent:
        print(f"Downloading {config.name}...")
    try:
        return await _download_tool(tool)
    except Exception as error:
        if not silent:
            print(f"Failed to download {config.name}: {error}")
        return None

File: packages/coding-agent/src/utils/tools-manager.ts
import { createHash } from "node:crypto";
import chalk from "chalk";
import { type SpawnSyncReturns, spawnSync } from "child_process";
import {
	chmodSync,
	createReadStream,
	createWriteStream,
	existsSync,
	mkdirSync,
	readdirSync,
	renameSync,
	rmSync,
} from "fs";
import { arch, platform } from "os";
import { join } from "path";
import { Readable } from "stream";
import { pipeline } from "stream/promises";
import { APP_NAME, getBinDir } from "../config.ts";

const TOOLS_DIR = getBinDir();
const NETWORK_TIMEOUT_MS = 10_000;
const DOWNLOAD_TIMEOUT_MS = 120_000;

function isOfflineModeEnabled(): boolean {
	const value = process.env.PI_OFFLINE;
	if (!value) return false;
	return value === "1" || value.toLowerCase() === "true" || value.toLowerCase() === "yes";
}

interface ToolConfig {
	name: string;
	repo: string; // GitHub repo (e.g., "sharkdp/fd")
	binaryName: string; // Name of the binary inside the archive
	systemBinaryNames?: string[]; // Alternative system command names to try before downloading
	tagPrefix: string; // Prefix for tags (e.g., "v" for v1.0.0, "" for 1.0.0)
	getAssetName: (version: string, plat: string, architecture: string) => string | null;
	/** Fixed version to download instead of resolving "latest" (reproducibility). */
	pinnedVersion?: string;
	/** SHA-256 of the release archive per `${plat}/${architecture}` (verification). */
	checksums?: Record<string, string>;
	/** If the release ships `<asset><checksumSuffix>` checksum files, verify against them. */
	checksumSuffix?: string;
}

const TOOLS: Record<string, ToolConfig> = {
	fd: {
		name: "fd",
		repo: "sharkdp/fd",
		binaryName: "fd",
		systemBinaryNames: ["fd", "fdfind"],
		tagPrefix: "v",
		// fd dropped x86_64 darwin builds after 10.3.0, so the version is pinned to
		// a release that covers every platform with a known SHA-256.
		pinnedVersion: "10.3.0",
		checksums: {
			"darwin/aarch64": "0570263812089120bc2a5d84f9e65cd0c25e4a4d724c80075c357239c74ae904",
			"darwin/x64": "50d30f13fe3d5914b14c4fff5abcbd4d0cdab4b855970a6956f4f006c17117a3",
			"linux/aarch64": "66f297e404400a3358e9a0c0b2f3f4725956e7e4435427a9ae56e22adbe73a68",
			"linux/x64": "c3c2bc79f838e780173fc8f18b337ec273e7ba17c7ff8f551be29fc3c19b7916",
			"win32/aarch64": "bf9b1e31bcac71c1e95d49c56f0d872f525b95d03854e94b1d4dd6786f825cc5",
			"win32/x64": "318aa2a6fa664325933e81fda60d523fff29444129e91ebf0726b5b3bcd8b059",
		},
		getAssetName: (version, plat, architecture) => {
			if (plat === "darwin") {
				const archStr = architecture === "arm64" ? "aarch64" : "x86_64";
				return `fd-v${version}-${archStr}-apple-darwin.tar.gz`;
			} else if (plat === "linux") {
				const archStr = architecture === "arm64" ? "aarch64" : "x86_64";
				return `fd-v${version}-${archStr}-unknown-linux-gnu.tar.gz`;
			} else if (plat === "win32") {
				const archStr = architecture === "arm64" ? "aarch64" : "x86_64";
				return `fd-v${version}-${archStr}-pc-windows-msvc.zip`;
			}
			return null;
		},
	},
	rg: {
		name: "ripgrep",
		repo: "BurntSushi/ripgrep",
		binaryName: "rg",
		tagPrefix: "",
		// ripgrep ships a `<asset>.sha256` file next to every release asset.
		checksumSuffix: ".sha256",
		getAssetName: (version, plat, architecture) => {
			if (plat === "darwin") {
				const archStr = architecture === "arm64" ? "aarch64" : "x86_64";
				return `ripgrep-${version}-${archStr}-apple-darwin.tar.gz`;
			} else if (plat === "linux") {
				if (architecture === "arm64") {
					return `ripgrep-${version}-aarch64-unknown-linux-gnu.tar.gz`;
				}
				return `ripgrep-${version}-x86_64-unknown-linux-musl.tar.gz`;
			} else if (plat === "win32") {
				const archStr = architecture === "arm64" ? "aarch64" : "x86_64";
				return `ripgrep-${version}-${archStr}-pc-windows-msvc.zip`;
			}
			return null;
		},
	},
};

// Check if a command exists in PATH by trying to run it
function commandExists(cmd: string): boolean {
	try {
		const result = spawnSync(cmd, ["--version"], { stdio: "pipe" });
		// Check for ENOENT error (command not found)
		return result.error === undefined || result.error === null;
	} catch {
		return false;
	}
}

// Get the path to a tool (system-wide or in our tools dir)
export function getToolPath(tool: "fd" | "rg"): string | null {
	const config = TOOLS[tool];
	if (!config) return null;

	// Check our tools directory first
	const localPath = join(TOOLS_DIR, config.binaryName + (platform() === "win32" ? ".exe" : ""));
	if (existsSync(localPath)) {
		return localPath;
	}

	// Check system PATH - if found, just return the command name (it's in PATH)
	const systemBinaryNames = config.systemBinaryNames ?? [config.binaryName];
	for (const systemBinaryName of systemBinaryNames) {
		if (commandExists(systemBinaryName)) {
			return systemBinaryName;
		}
	}

	return null;
}

// Fetch latest release version from GitHub
async function getLatestVersion(repo: string): Promise<string> {
	const response = await fetch(`https://api.github.com/repos/${repo}/releases/latest`, {
		headers: { "User-Agent": `${APP_NAME}-coding-agent` },
		signal: AbortSignal.timeout(NETWORK_TIMEOUT_MS),
	});

	if (!response.ok) {
		throw new Error(`GitHub API error: ${response.status}`);
	}

	const data = (await response.json()) as { tag_name: string };
	return data.tag_name.replace(/^v/, "");
}

// Download a file from URL
async function downloadFile(url: string, dest: string): Promise<void> {
	const response = await fetch(url, {
		signal: AbortSignal.timeout(DOWNLOAD_TIMEOUT_MS),
	});

	if (!response.ok) {
		throw new Error(`Failed to download: ${response.status}`);
	}
	if (!response.body) {
		throw new Error("No response body");
	}

	const fileStream = createWriteStream(dest);
	await pipeline(Readable.fromWeb(response.body as any), fileStream);
}

async function verifySha256(filePath: string, expectedHex: string, assetName: string): Promise<void> {
	const hash = createHash("sha256");
	await new Promise<void>((resolve, reject) => {
		const stream = createReadStream(filePath);
		stream.on("data", (chunk) => hash.update(chunk));
		stream.on("end", () => resolve());
		stream.on("error", reject);
	});
	const actualHex = hash.digest("hex");
	if (actualHex !== expectedHex.toLowerCase()) {
		throw new Error(`Checksum mismatch for ${assetName}: expected ${expectedHex}, got ${actualHex}`);
	}
}

function findBinaryRecursively(rootDir: string, binaryFileName: string): string | null {
	const stack: string[] = [rootDir];

	while (stack.length > 0) {
		const currentDir = stack.pop();
		if (!currentDir) continue;

		const entries = readdirSync(currentDir, { withFileTypes: true });
		for (const entry of entries) {
			const fullPath = join(currentDir, entry.name);
			if (entry.isFile() && entry.name === binaryFileName) {
				return fullPath;
			}
			if (entry.isDirectory()) {
				stack.push(fullPath);
			}
		}
	}

	return null;
}

function formatSpawnFailure(result: SpawnSyncReturns<Buffer>): string {
	if (result.error?.message) {
		return result.error.message;
	}
	const stderr = result.stderr?.toString().trim();
	if (stderr) {
		return stderr;
	}
	const stdout = result.stdout?.toString().trim();
	if (stdout) {
		return stdout;
	}
	return `exit status ${result.status ?? "unknown"}`;
}

function runExtractionCommand(command: string, args: string[]): string | null {
	const result = spawnSync(command, args, { stdio: "pipe" });
	if (!result.error && result.status === 0) {
		return null;
	}
	return `${command}: ${formatSpawnFailure(result)}`;
}

function extractTarGzArchive(archivePath: string, extractDir: string, assetName: string): void {
	const failure = runExtractionCommand("tar", ["xzf", archivePath, "-C", extractDir]);
	if (failure) {
		throw new Error(`Failed to extract ${assetName}: ${failure}`);
	}
}

function getWindowsTarCommand(): string {
	const systemRoot = process.env.SystemRoot ?? process.env.WINDIR;
	if (systemRoot) {
		const systemTar = join(systemRoot, "System32", "tar.exe");
		if (existsSync(systemTar)) {
			return systemTar;
		}
	}
	return "tar.exe";
}

function extractZipArchive(archivePath: string, extractDir: string, assetName: string): void {
	const failures: string[] = [];

	if (platform() === "win32") {
		// Windows ships bsdtar as tar.exe, which supports zip files. Prefer the
		// System32 binary over Git Bash's GNU tar, which does not handle zip archives.
		const tarFailure = runExtractionCommand(getWindowsTarCommand(), ["xf", archivePath, "-C", extractDir]);
		if (!tarFailure) return;
		failures.push(tarFailure);

		const script =
			"& { param($archive, $destination) $ErrorActionPreference = 'Stop'; Expand-Archive -LiteralPath $archive -DestinationPath $destination -Force }";
		const powershellFailure = runExtractionCommand("powershell.exe", [
			"-NoLogo",
			"-NoProfile",
			"-NonInteractive",
			"-ExecutionPolicy",
			"Bypass",
			"-Command",
			script,
			archivePath,
			extractDir,
		]);
		if (!powershellFailure) return;
		failures.push(powershellFailure);
	} else {
		const unzipFailure = runExtractionCommand("unzip", ["-q", archivePath, "-d", extractDir]);
		if (!unzipFailure) return;
		failures.push(unzipFailure);

		const tarFailure = runExtractionCommand("tar", ["xf", archivePath, "-C", extractDir]);
		if (!tarFailure) return;
		failures.push(tarFailure);
	}

	throw new Error(`Failed to extract ${assetName}: ${failures.join("; ")}`);
}

// Download and install a tool
async function downloadTool(tool: "fd" | "rg"): Promise<string> {
	const config = TOOLS[tool];
	if (!config) throw new Error(`Unknown tool: ${tool}`);

	const plat = platform();
	const architecture = arch();

	// Get latest version (or the pinned one for reproducible downloads)
	const version = config.pinnedVersion ?? (await getLatestVersion(config.repo));

	// Get asset name for this platform
	const assetName = config.getAssetName(version, plat, architecture);
	if (!assetName) {
		throw new Error(`Unsupported platform: ${plat}/${architecture}`);
	}

	// Create tools directory
	mkdirSync(TOOLS_DIR, { recursive: true });

	const downloadUrl = `https://github.com/${config.repo}/releases/download/${config.tagPrefix}${version}/${assetName}`;
	const archivePath = join(TOOLS_DIR, assetName);
	const binaryExt = plat === "win32" ? ".exe" : "";
	const binaryPath = join(TOOLS_DIR, config.binaryName + binaryExt);

	// Download
	await downloadFile(downloadUrl, archivePath);

	// Verify the archive's SHA-256 before extraction so a tampered or
	// misdirected download is never executed.
	if (config.checksums) {
		const expectedHex = config.checksums[`${plat}/${architecture}`];
		if (!expectedHex) {
			rmSync(archivePath, { force: true });
			throw new Error(`No checksum configured for ${tool} on ${plat}/${architecture}`);
		}
		await verifySha256(archivePath, expectedHex, assetName);
	} else if (config.checksumSuffix) {
		const checksumUrl = `${downloadUrl}${config.checksumSuffix}`;
		const checksumResponse = await fetch(checksumUrl, {
			signal: AbortSignal.timeout(DOWNLOAD_TIMEOUT_MS),
		});
		if (!checksumResponse.ok) {
			rmSync(archivePath, { force: true });
			throw new Error(`Failed to fetch checksum (${checksumResponse.status}): ${checksumUrl}`);
		}
		const expectedHex = (await checksumResponse.text()).trim().split(/\s+/)[0];
		if (!expectedHex) {
			rmSync(archivePath, { force: true });
			throw new Error(`Empty checksum from ${checksumUrl}`);
		}
		await verifySha256(archivePath, expectedHex, assetName);
	}

	// Extract into a unique temp directory. fd and rg downloads can run concurrently
	// during startup, so sharing a fixed directory causes races.
	const extractDir = join(
		TOOLS_DIR,
		`extract_tmp_${config.binaryName}_${process.pid}_${Date.now()}_${Math.random().toString(36).slice(2, 10)}`,
	);
	mkdirSync(extractDir, { recursive: true });

	try {
		if (assetName.endsWith(".tar.gz")) {
			extractTarGzArchive(archivePath, extractDir, assetName);
		} else if (assetName.endsWith(".zip")) {
			extractZipArchive(archivePath, extractDir, assetName);
		} else {
			throw new Error(`Unsupported archive format: ${assetName}`);
		}

		// Find the binary in extracted files. Some archives contain files directly
		// at root, others nest under a versioned subdirectory.
		const binaryFileName = config.binaryName + binaryExt;
		const extractedDir = join(extractDir, assetName.replace(/\.(tar\.gz|zip)$/, ""));
		const extractedBinaryCandidates = [join(extractedDir, binaryFileName), join(extractDir, binaryFileName)];
		let extractedBinary = extractedBinaryCandidates.find((candidate) => existsSync(candidate));

		if (!extractedBinary) {
			extractedBinary = findBinaryRecursively(extractDir, binaryFileName) ?? undefined;
		}

		if (extractedBinary) {
			renameSync(extractedBinary, binaryPath);
		} else {
			throw new Error(`Binary not found in archive: expected ${binaryFileName} under ${extractDir}`);
		}

		// Make executable (Unix only)
		if (plat !== "win32") {
			chmodSync(binaryPath, 0o755);
		}
	} finally {
		// Cleanup
		rmSync(archivePath, { force: true });
		rmSync(extractDir, { recursive: true, force: true });
	}

	return binaryPath;
}

// Termux package names for tools
const TERMUX_PACKAGES: Record<string, string> = {
	fd: "fd",
	rg: "ripgrep",
};

// Ensure a tool is available, downloading if necessary
// Returns the path to the tool, or null if unavailable
export async function ensureTool(tool: "fd" | "rg", silent: boolean = false): Promise<string | undefined> {
	const existingPath = getToolPath(tool);
	if (existingPath) {
		return existingPath;
	}

	const config = TOOLS[tool];
	if (!config) return undefined;

	if (isOfflineModeEnabled()) {
		if (!silent) {
			console.log(chalk.yellow(`${config.name} not found. Offline mode enabled, skipping download.`));
		}
		return undefined;
	}

	// On Android/Termux, Linux binaries don't work due to Bionic libc incompatibility.
	// Users must install via pkg.
	if (platform() === "android") {
		const pkgName = TERMUX_PACKAGES[tool] ?? tool;
		if (!silent) {
			console.log(chalk.yellow(`${config.name} not found. Install with: pkg install ${pkgName}`));
		}
		return undefined;
	}

	// Tool not found - download it
	if (!silent) {
		console.log(chalk.dim(`${config.name} not found. Downloading...`));
	}

	try {
		const path = await downloadTool(tool);
		if (!silent) {
			console.log(chalk.dim(`${config.name} installed to ${path}`));
		}
		return path;
	} catch (e) {
		if (!silent) {
			console.log(chalk.yellow(`Failed to download ${config.name}: ${e instanceof Error ? e.message : e}`));
		}
		return undefined;
	}
}
